lib/smack/SplitStructLoadStore.cpp

```diff
@@ -0,0 +1,148 @@
+#include "smack/SplitStructLoadStore.h"
+#include "llvm/IR/InstIterator.h"
+
+namespace smack {
+
+using namespace llvm;
+
+std::vector<Value*> getFirsts(std::vector<std::pair<Value*, unsigned> > lst)
+{
+    std::vector<Value*> ret;
+    for (auto p = lst.begin(); p != lst.end(); ++p)
+      ret.push_back(std::get<0>(*p));
+    return ret;
+}
+
+std::vector<unsigned> getSeconds(std::vector<std::pair<Value*, unsigned> > lst)
+{
+    std::vector<unsigned> ret;
+    for (auto p = lst.begin()+1; p != lst.end(); ++p)
+      ret.push_back(std::get<1>(*p));
+    return ret;
+}
+
+bool SplitStructLoadStore::runOnBasicBlock(BasicBlock& BB)
+{
+  std::vector<Instruction*> toRemove;
+  for(Instruction& I : BB) {
+    if (LoadInst* li = dyn_cast<LoadInst>(&I)) {
+      bool isReturnValue = false;
+      for (auto u = li->user_begin(); u != li->user_end(); ++u)
+      {
+        if (auto ri = dyn_cast<ReturnInst>(*u)) {
+          isReturnValue = true;
+          break;
+        }
+      }
+      if (!isReturnValue && li->getType()->isAggregateType()) {
+        splitStructLoad(li);
+        toRemove.push_back(li);
+      }
+    } else if (StoreInst* si = dyn_cast<StoreInst>(&I)) {
+      Value* P = si->getPointerOperand();
+      Value* V = si->getOperand(0)->stripPointerCasts();
+      if (V->getType()->isAggregateType()) {
+        splitStructStore(si, P, V);
+        toRemove.push_back(si);
+      }
+    }
+  }
+
+  for (auto& i : toRemove)
+    i->eraseFromParent();
+  return true;
+}
+
+void SplitStructLoadStore::splitStructLoad(LoadInst* li)
+{
+  if (StructType* st = dyn_cast<StructType>(li->getType())) {
+    std::vector<std::pair<Value*, unsigned> > idx;
+    IRBuilder<> irb(li);
+    li->replaceAllUsesWith(buildStructs(&irb, li->getPointerOperand(), st, nullptr, idx));
+  }
+}
+
+Value* SplitStructLoadStore::buildStructs(IRBuilder<> *irb, Value* ptr, Type* ct, Value* val, std::vector<std::pair<Value*, unsigned> > idxs)
+{
+  LLVMContext& C = ptr->getContext();
+  Value* cv = val? val : UndefValue::get(ct);
+
+  if (ct->isIntegerTy() || ct->isPointerTy() || ct->isFloatingPointTy())
+    return irb->CreateInsertValue(val,
+             irb->CreateLoad(irb->CreateGEP(ptr, ArrayRef<Value*>(getFirsts(idxs)))),
```

Comment:
I think the way that this line is split makes sense to me because each argument of the top level function call is put into separate lines. It should help readability.